tensorboard/uploader/uploader_main.py

```diff
@@ -348,11 +358,12 @@ def execute(self, server_info, channel):
         )
         gen = exporter_lib.list_experiments(api_client, fieldmask=fieldmask)
         count = 0
+        if self.json:
```

Comment:
The code paths here are asymmetric and intertwined (as the need for multiple `if self.json:` shows).  This is in part because the json path collects all the entries first and then prints them at the end (as noted in the PR description), but the non-json version is streaming.

First let's make this print exactly the same json output, except in a streaming manner:

```
if self.json:
    print "["
for experiment in gen:
    ...
    if self.json:
        print(textwrap.indent(json.dumps(data, indent=2), prefix="  "))
    ...
if self.json:
    print "]"
```

Next, to be more persnickety and to make it easier to add other formats, you could factor out the format-specific parts by creating an abstract `Formatter` class with methods `prefix()`, `format_experiment(data)`, and `suffix()`.  Then implement those for text and json, respectively, and pass the appropriate Formatter in to `_ListIntent` (or at least, pick the right one in the constructor).

By making the Formatter methods return strings (as opposed to `print(...)`ing as a side effect) you can make the choice of destination (stdout vs file, etc.) an orthogonal concern.

Reply:
Awesome, thanks!